provider/k8s/client.go

```diff
@@ -227,10 +239,15 @@ func (c *clientImpl) watch(url string, stopCh <-chan bool) (chan interface{}, ch
 		return watchCh, errCh, fmt.Errorf("failed to decode version %v", err)
 	}
 	resourceVersion := generic.ResourceVersion
-
-	url = url + "?watch&resourceVersion=" + resourceVersion
-	// Make request to Kubernetes API
-	request := c.request(url)
+	queryParams := map[string]string{"watch": "", "resourceVersion": resourceVersion}
+	if labelSelector != "" {
+		queryParams["labelSelector"] = labelSelector
+	}
+	queryData, err := json.Marshal(queryParams)
+	if err != nil {
+		return watchCh, errCh, fmt.Errorf("Unable to construct query args")
+	}
+	request := c.request(url, string(queryData))
```

Comment:
This looks very similar to L57-65 could we extract a function to dry it up a little.